tests/slist_ut/slist_ut.c

```diff
@@ -0,0 +1,1114 @@
+TEST_FUNCTION(slist_is_empty_with_one_entry_succeeds)
+{
+    // arrange
+    static simpleItem simp1 = { 1, { NULL } };
+
+    SINGLYLINKEDLIST_ENTRY head;
+    slist_initialize(&head);
+
+    slist_add(&head, &(simp1.link));
+
+    // act
+    bool result = slist_is_empty(&head);
+
+    // assert
+    ASSERT_ARE_EQUAL(bool, false, result);
+}
+
+/* Tests_SRS_SLIST_07_004: [ `slist_is_empty` shall return `false` if there is one or more entris in the list. ]*/
+TEST_FUNCTION(slist_is_empty_with_multiple_entries_succeeds)
+{
+    // arrange
+    static simpleItem simp1 = { 1, { NULL } };
+    static simpleItem simp2 = { 2, { NULL } };
+    static simpleItem simp3 = { 3, { NULL } };
+
+    SINGLYLINKEDLIST_ENTRY head;
+    slist_initialize(&head);
+    slist_add(&head, &(simp1.link));
+    slist_add(&head, &(simp2.link));
+    slist_add(&head, &(simp3.link));
+
+    // act
+    bool result = slist_is_empty(&head);
+
+    // assert
+    ASSERT_ARE_EQUAL(bool, false, result);
+}
+
+/*slist_add*/
+
+/* Tests_SRS_SLIST_07_005: [ If `list_head` is `NULL`, `slist_add` shall fail and return `NULL`. ]*/
+TEST_FUNCTION(slist_add_with_head_entry_NULL_fails)
+{
+    // arrange
+    static simpleItem simp1 = { 1, { NULL } };
+
+    // act
+    PSINGLYLINKEDLIST_ENTRY result = slist_add(NULL, &(simp1.link));
+
+    // assert
+    ASSERT_ARE_EQUAL(void_ptr, NULL, result);
+}
+
+/* Tests_SRS_SLIST_07_006: [ If `list_entry` is `NULL`, `slist_add` shall fail and return `NULL`. ]*/
+TEST_FUNCTION(slist_add_with_add_entry_NULL_fails)
+{
+    // arrange
+    SINGLYLINKEDLIST_ENTRY head;
+    slist_initialize(&head);
+
+    // act
+    PSINGLYLINKEDLIST_ENTRY result = slist_add(&head, NULL);
+
+    // assert
+    ASSERT_ARE_EQUAL(void_ptr, NULL, result);
+}
+
+/* Tests_SRS_SLIST_07_007: [ `slist_add` shall add one entry to the tail of the list on success and return a pointer to the added entry. ]*/
+TEST_FUNCTION(slist_add_with_empty_list_succeeds)
+{
+    // arrange
+    static simpleItem simp1 = { 1, { NULL } };
+    SINGLYLINKEDLIST_ENTRY head;
+    slist_initialize(&head);
+ 
+    // act
+    PSINGLYLINKEDLIST_ENTRY result = slist_add(&head, &(simp1.link));
+
+    // assert
+    ASSERT_IS_NOT_NULL(result);
+    ASSERT_ARE_EQUAL(void_ptr, &(simp1.link), result);
+    ASSERT_ARE_EQUAL(void_ptr, head.next, result);
+}
+
+/* Tests_SRS_SLIST_07_007: [ `slist_add` shall add one entry to the tail of the list on success and return a pointer to the added entry. ]*/
+TEST_FUNCTION(slist_add_with_multiple_entries_in_the_list_succeeds)
+{
+    // arrange
+    static simpleItem simp1 = { 1, { NULL } };
+    static simpleItem simp2 = { 2, { NULL } };
+    static simpleItem simp3 = { 3, { NULL } };
+    static simpleItem simp4 = { 4, { NULL } };
+
+    SINGLYLINKEDLIST_ENTRY head;
+    slist_initialize(&head);
+    head.next = &(simp1.link);
+    head.next->next = &(simp2.link);
+    head.next->next->next = &(simp3.link);
```

Comment:
We should not be manipulating these manually. We should build the list using the APIs we expose.

Reply:
Here and in all other places.

Reply:
done